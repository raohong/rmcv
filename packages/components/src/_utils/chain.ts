const chain = <T extends (...args: any[]) => void>(
  ...fns: (T | undefined)[]
): ((...args: Parameters<T>) => void) => {
  return (...args) => {
    fns.forEach((item) => {
      if (item) {
        item(...args);
      }
    });
  };
};

export default chain;
