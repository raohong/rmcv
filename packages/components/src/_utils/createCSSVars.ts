import kebabCase from 'lodash/kebabCase';

const createCSSVars = (
  prefix: string,
  style: Record<string, string | undefined | null>,
) => {
  const result = Object.entries(style).reduce((map, [key, value]) => {
    if (!value) {
      return map;
    }
    // @ts-ignore
    // eslint-disable-next-line no-param-reassign
    map[`--${prefix}-${kebabCase(key)}`] = value;

    return map;
  }, {} as React.CSSProperties);

  return result;
};

export default createCSSVars;
