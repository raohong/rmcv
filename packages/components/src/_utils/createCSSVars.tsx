import React from 'react';

export const createCSSVars = (vars: Record<string, number | string>) =>
  Object.entries(vars).reduce((result, [name, value]) => {
    // @ts-ignore
    result[`--${name}`] = value;
    return result;
  }, {} as React.CSSProperties);
