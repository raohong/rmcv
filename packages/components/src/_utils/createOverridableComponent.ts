import type {
  ComponentClass,
  ComponentProps,
  ComponentType,
  FC,
  ForwardRefExoticComponent,
  JSXElementConstructor,
  PropsWithoutRef,
  RefAttributes,
} from 'react';

type Union<Base extends object, Target extends object> = Pick<
  Target,
  Exclude<keyof Target, keyof Base>
> &
Base;

export interface OverridableComponent<
  C extends ComponentType<ComponentProps<D>>,
  DefaultElement extends keyof JSX.IntrinsicElements = 'div',
  AdditionalProps extends object = object,
  ExcludeProps extends object = object,
  D extends keyof JSX.IntrinsicElements | React.JSXElementConstructor<any> = | keyof JSX.IntrinsicElements
  | React.JSXElementConstructor<any>,
> extends FC<
  Omit<
    Union<ComponentProps<C>, JSX.IntrinsicElements[DefaultElement]>,
      keyof ExcludeProps | 'component'
  > &
  AdditionalProps
  > {
  // JSX.IntrinsicElements
  <Tag extends keyof JSX.IntrinsicElements>(
    props: Omit<
      Union<PropsWithoutRef<ComponentProps<C>>, JSX.IntrinsicElements[Tag]>,
      keyof ExcludeProps | 'component'
    > & {
      /**
       * @description 自定义 component
       * @default div
       */
      component: Tag;
    } & AdditionalProps,
  ): JSX.Element;

  // ClassComponent
  <
    Custom extends ComponentClass<ComponentProps<D>>,
    D extends keyof JSX.IntrinsicElements | JSXElementConstructor<any> = | keyof JSX.IntrinsicElements
    | JSXElementConstructor<any>,
  >(
    props: Omit<
      Union<PropsWithoutRef<ComponentProps<C>>, ComponentProps<Custom>>,
      keyof ExcludeProps | 'component'
    > & {
      /**
       * @description 自定义 component
       */
      component: Custom;
    } & RefAttributes<InstanceType<Custom>> &
    AdditionalProps,
  ): JSX.Element;

  <
    Custom extends ForwardRefExoticComponent<ComponentProps<D>>,
    D extends keyof JSX.IntrinsicElements | JSXElementConstructor<any> = | keyof JSX.IntrinsicElements
    | JSXElementConstructor<any>,
  >(
    props: Omit<
      Union<PropsWithoutRef<ComponentProps<C>>, ComponentProps<Custom>>,
      keyof ExcludeProps | 'component'
    > & {
      /**
       * @description 自定义 component
       */
      component: Custom;
    } & AdditionalProps,
  ): JSX.Element;
}

export function createOverridableComponent<
  C extends ComponentType<ComponentProps<C>>,
  DefaultElement extends keyof JSX.IntrinsicElements = 'div',
  AdditionalProps extends object = object,
  ExcludeProps extends object = object,
>(
  component: C,
): OverridableComponent<C, DefaultElement, AdditionalProps, ExcludeProps> {
  return component as unknown as OverridableComponent<
    C,
    DefaultElement,
    AdditionalProps,
    ExcludeProps
  >;
}
