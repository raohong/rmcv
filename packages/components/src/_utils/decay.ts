import { useEffect, useRef, useCallback } from 'react';
import type { SpringValue } from '@react-spring/web';
import raf from 'raf';

export type DecayAnimationOptions = {
  power?: number;
  timeConst?: number;
  from?: number;
  velocity?: number;
  modifyTarget?: (dist: number, velocity: number) => number;
  restDelta?: number;
};

type DecayAnimationState = {
  finished: boolean;
  value: number;
  velocity: number;
};

const decay = ({
  power = 0.8,
  timeConst = 300,
  from = 0,
  velocity = 0,
  restDelta = 0.2,
  modifyTarget,
}: DecayAnimationOptions) => {
  let movement = velocity * power * timeConst;
  const ideal = movement + from;

  const target = modifyTarget ? modifyTarget(ideal, velocity) : ideal;
  const state: DecayAnimationState = {
    finished: false,
    value: from,
    velocity,
  };

  if (target !== ideal) {
    movement = target - from;
  }

  return {
    next: (t: number) => {
      // y = a * Math.exp(kt)

      const dx = Math.exp((1 / -timeConst) * t);
      const delta = -movement * dx;

      state.finished = Math.abs(delta) < restDelta;
      state.value = state.finished ? target : target + delta;
      state.velocity = velocity * dx;

      return state;
    },
  };
};

export const runDecayAnimation = ({
  onChange,
  ...options
}: DecayAnimationOptions & {
  onChange: (state: DecayAnimationState) => void;
}) => {
  let id: number | null = null;
  let active = true;
  const startTime = Date.now();

  const generator = decay(options);

  const cancel = () => {
    if (id !== null) {
      raf.cancel(id);
      id = null;
    }

    active = false;
  };

  const run = () => {
    id = raf(() => {
      if (active) {
        const now = Date.now();
        const state = generator.next(now - startTime);

        onChange(state);

        if (!state.finished) {
          run();
        }
      }
    });
  };

  run();

  return cancel;
};

export const useDecayAnimation = () => {
  const unmountedRef = useRef(false);
  const canceller = useRef<(() => void)[]>([]);

  const cancelAnimation = useCallback(() => {
    if (canceller.current) {
      canceller.current.forEach((item) => item());
    }
  }, []);

  const runDecay = useCallback(
    (
      spring: SpringValue<number>,
      dir: number,
      boundary: number,
      options: DecayAnimationOptions,
    ) => {
      const cb = runDecayAnimation({
        ...options,
        onChange({ velocity, value }) {
          if (unmountedRef.current) {
            return;
          }

          if (
            (dir === 1 && value >= boundary) ||
            (dir === -1 && value <= boundary)
          ) {
            cb();
            canceller.current = canceller.current.filter((item) => item !== cb);

            spring.start({
              to: boundary,
              config: {
                decay: false,
                velocity,
              },
            });
          } else {
            spring.set(value);
          }
        },
      });

      canceller.current.push(cb);
    },
    [],
  );

  useEffect(() => {
    return () => {
      cancelAnimation();
    };
  }, [cancelAnimation]);

  return { runDecay, cancelAnimation };
};
