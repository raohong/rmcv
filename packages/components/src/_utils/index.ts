export { default as getPrefixCls } from './getPrefixCls';

export { default as isBrowser } from './isBrowser';

export { default as renderPortal } from './renderPortal';

export { default as uuid } from './uuid';

export { default as createCSSVars } from './createCSSVars';

export { default as chain } from './chain';

export { default as toArray } from './toArray';
