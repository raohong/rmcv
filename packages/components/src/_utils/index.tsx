export * from './createOverridableComponent';
export * from './decay';
export * from './flatReactNode';
export * from './getPrefixCls';
export * from './htmlAttrs';
export * from './reactRender';
export * from './useXId';
