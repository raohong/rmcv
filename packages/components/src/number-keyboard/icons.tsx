import React from 'react';

export const DeleteIcon: React.FC = () => (
  <svg viewBox="0 0 32 22">
    <path
      d="M28 0a4 4 0 0 1 4 4v14a4 4 0 0 1-4 4H10.4a2 2 0 0 1-1.4-.6L1 13.1c-.6-.5-.9-1.3-.9-2 0-1 .3-1.7.9-2.2L9 .6a2 2 0 0 1 1.4-.6zm0 2H10.4l-8.2 8.3a1 1 0 0 0-.3.7c0 .3.1.5.3.7l8.2 8.4H28a2 2 0 0 0 2-2V4c0-1.1-.9-2-2-2zm-5 4a1 1 0 0 1 .7.3 1 1 0 0 1 0 1.4L20.4 11l3.3 3.3c.2.2.3.5.3.7 0 .3-.1.5-.3.7a1 1 0 0 1-.7.3 1 1 0 0 1-.7-.3L19 12.4l-3.4 3.3a1 1 0 0 1-.6.3 1 1 0 0 1-.7-.3 1 1 0 0 1-.3-.7c0-.2.1-.5.3-.7l3.3-3.3-3.3-3.3A1 1 0 0 1 14 7c0-.3.1-.5.3-.7A1 1 0 0 1 15 6a1 1 0 0 1 .6.3L19 9.6l3.3-3.3A1 1 0 0 1 23 6z"
      fill="currentColor"
    ></path>
  </svg>
);

export const CollapseIcon: React.FC = () => (
  <svg viewBox="0 0 30 24">
    <path
      d="M26 13h-2v2h2v-2zm-8-3h2V8h-2v2zm2-4h2V4h-2v2zm2 4h4V4h-2v4h-2v2zm-7 14 3-3h-6l3 3zM6 13H4v2h2v-2zm16 0H8v2h14v-2zm-12-3h2V8h-2v2zM28 0l1 1 1 1v15l-1 2H1l-1-2V2l1-1 1-1zm0 2H2v15h26V2zM6 4v2H4V4zm10 2h2V4h-2v2zM8 9v1H4V8zm8 0v1h-2V8zm-6-5v2H8V4zm4 0v2h-2V4z"
      fill="currentColor"
    ></path>
  </svg>
);
