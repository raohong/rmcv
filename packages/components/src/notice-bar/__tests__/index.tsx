import { VolumeOutlined } from '@rmc-vant/icons';
import { fireEvent, render, screen } from '@testing-library/react';
import { NoticeBar } from '..';
import { getPrefixCls } from '../../_utils';

const testId = 'notice-bar';

it('render correctly', () => {
  const tree = render(<NoticeBar text='text' />);

  expect(tree.asFragment()).toMatchSnapshot();
});

it('render with color', () => {
  render(<NoticeBar color='red' data-testid={testId} />);

  expect(screen.getByTestId(testId)).toHaveStyle({
    color: 'red',
  });
});

it('render with background', () => {
  render(<NoticeBar background='blue' data-testid={testId} />);

  expect(screen.getByTestId(testId)).toHaveStyle({
    background: 'blue',
  });
});

it('render with iconColor and leftColor', () => {
  render(
    <NoticeBar
      leftIcon={<VolumeOutlined data-testid='volume-icon' />}
      iconColor='green'
      data-testid={testId}
    />,
  );

  const icons = screen
    .getByTestId(testId)
    .querySelectorAll(`.${getPrefixCls('notice-bar-icon')}`);

  expect(screen.getByTestId('volume-icon')).toBeInTheDocument();

  icons.forEach((item) => {
    expect(item).toHaveStyle({
      color: 'green',
    });
  });
});

it('render with mode: closeable', () => {
  const onClose = jest.fn();

  render(<NoticeBar mode='closeable' onClose={onClose} data-testid={testId} />);

  expect(screen.getByLabelText('Cross')).toBeInTheDocument();

  fireEvent.click(screen.getByLabelText('Cross'));

  expect(onClose).toHaveBeenCalled();
});

it('render with mode: link', () => {
  render(<NoticeBar mode='link' data-testid={testId} />);

  expect(screen.getByLabelText('Arrow')).toBeInTheDocument();
});
