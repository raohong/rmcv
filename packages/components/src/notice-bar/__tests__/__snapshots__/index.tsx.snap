// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render correctly 1`] = `
<DocumentFragment>
  .emotion-0 {
  box-sizing: border-box;
  margin: 0;
  padding: 0px 16px;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-tap-highlight-color: transparent;
  -webkit-font-smooth: antialiased;
  position: relative;
  touch-action: none;
  list-style: none;
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  font-family: Open Sans,-apple-system,BlinkMacSystemFont,Helvetica Neue,Helvetica,Segoe UI,Arial,Roboto,PingFang SC,miui,Hiragino Sans GB,Microsoft Yahei,sans-serif;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-size: 14px;
  line-height: 24px;
  color: #ed6a0c;
  background: #fffbe8;
  height: 40px;
  white-space: nowrap;
}

.emotion-1 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  overflow: hidden;
  box-sizing: border-box;
}

.emotion-2 {
  box-sizing: border-box;
  overflow: hidden;
  text-overflow: ellipsis;
}

<div
    class="RVuiNoticeBar-root emotion-0"
  >
    <div
      class="emotion-1"
    >
      <div
        class="RVuiNoticeBar-content emotion-2"
      >
        text
      </div>
    </div>
  </div>
</DocumentFragment>
`;
